### PR TITLE
Plugin/Metrics - add Healther interface and refactor uniqAddr storage

### PR DESCRIPTION
**Pull request by fturib**
### 1. Why is this pull request needed and what does it do?
Inform when Prometheus cannot provide service because HTTP server could not start.

NOTE: during testing, I found that the Tests could not run properly because the variable uniqAddr that is shared by all instance of metrics plugin was not reset  between Tests.
I moved that variable from global scope of "metrics" to the storage of each caddy instance.
(Caddy did that enhancement in purpose for the "reload" use case).

### Content
- register healther for metrics (return not ok if all listeners are not started properly)
- save the instance of uniqAddr, that is shared by all instances of metrics plugin, into the caddy's instance storage, instead global var.
- add UT with a sample of function RepairHealth for metrics

### 2. Which issues (if any) are related?
#1868. 
#1982 

#1868. is resolved by allowing a monitoring of the metrics using the health plugin. 
any monitor can then know when Prometheus does not provide the service and then decide to restart the process of CoreDNS.

#1982 is resolved because we create a new uniqAddr on every new instance : there is no more mismatch with setup of former instance.

### 3. Which documentation changes (if any) need to be made?
README of the plugin updated.

### NOTE: 
this PR overlaps with #1969 - the second to be merged will need a rebase before the merge.

### CODE REVIEW

File: plugin/pkg/uniq/uniq.go [hunk 1]
```diff
@@ -13,19 +18,23 @@ type item struct {
...
+	defer u.Unlock()
 	if _, ok := u.u[key]; ok {
 		return
 	}
 	u.u[key] = item{todo, f}
 }
 
 // SetTodo sets key to 'todo' again.
-func (u U) SetTodo(key string) {
+func (u *U) SetTodo(key string) {
```
Comment (miekg) [inline plugin/pkg/uniq/uniq.go]: Why did you change this to  a pointer receiver?

File: plugin/pkg/uniq/uniq.go [hunk 2]
```diff
@@ -34,18 +43,34 @@ func (u U) SetTodo(key string) {
...
 		}
-		v.state = done
-		u.u[k] = v
 	}
 	return nil
 }
 
+// HasTodo inform on weather some things are still in todo mode
+func (u *U) HasTodo() bool {
+	u.Lock()
```
Comment (miekg) [inline plugin/pkg/uniq/uniq.go]: This only needs a read lock

File: plugin/pkg/uniq/uniq.go (hunk 1 above)
Comment (fturib) [reply plugin/pkg/uniq/uniq.go]: Well, by setting the a value in U, I expect to modify that U structure.
It should then be a pointer receiver (instead of a value receiver).

I understand that was working so far. But I guess because the copy is sharing the same map.

But as I introduced the Mutex, this one was not effective if I use a value receiver : each copy of U was running on its own Mutex => I add a warning that mutex was inefficient and I could verify on UT that I had race conditions.

see : https://tour.golang.org/methods/4
https://programming.guide/go/pointer-vs-value-receiver.html

File: plugin/pkg/uniq/uniq.go (hunk 2 above)
Comment (fturib) [reply plugin/pkg/uniq/uniq.go]: ok.

File: plugin/pkg/uniq/uniq.go (hunk 1 above)
Comment (fturib) [reply plugin/pkg/uniq/uniq.go]: Right. a map varlue is a pointer to the head of map. So when copy we copy that pointer. Which make any change in that map works from the original or from the copy.

File: plugin/metrics/README.md [hunk 3]
```diff
@@ -39,6 +39,13 @@ name "dropped" (without a closing dot - this is never a valid domain name).
 
 This plugin can only be used once per Server Block.
 
+## Health
+
+This plugin implements dynamic health checking. 
+Prometheus plugin is considered unhealthy if the HTTP server instance was not successfully started.
```
Comment (miekg) [inline plugin/metrics/README.md:45]: so, we will not fail binding to the port? That was an earlier plan, no? Did we decide we could not do that ?

Comment (fturib) [reply plugin/metrics/README.md:45]: as you started to say in the issue #1868, the problem to fail is that if we fail to start the new server, the CoreDNS will continue service on the old server .. but the listener is in unknown state (most likely started to stop but not finished). Which means Metrics most likely stopped to serve information on either new or old server.

It does not help to automate the monitoring of CoreDNS, unless we completely stop CoreDNS.

Adding Metrics to Health and inform when metrics is unhealthy allow any monitor to take a decision for stop/start CoreDNS

**Comment by fturib** [comment]
@miekg : let me know if you can approve and merge here. This fix is blocking a push to production on our side. Thanks.!

File: plugin/metrics/README.md (hunk 3 above)
Comment (miekg) [reply plugin/metrics/README.md:45]: I keep finding this odd. We should probably crash the startup if we fail to bind, i.e. not fail the reload, crash the entire process. Seems simplest and avoid a whole bunch of code to work around it. I do wonder however if log.Fatalf would crash everything or just the reloading process.

I also don't understand why your monitoring doesn't see the monitoring endpoint disappearing and needs the health endpoint to do something

Comment (fturib) [reply plugin/metrics/README.md:45]: # Short

- **I can remove the healther implementation. I do not need it. But I would like to keep the move out of global var - which also fix #1868 **
- **I would not like to have to stop the process if Metrics cannot listen. At least for few weeks. That would block IB that try to go to production right now.**

# Longer reply 

I think this PR contains several changes, and maybe I should go one by one:

## 1- issue #1868
it is the impact of Reload on the Metrics plugin that is using uniq resources (the Listeners). 
IMO the root cause is the usage of global var to save information : we cannot cleanup this information during a reload. That get worse when the Reload fails for some reason and we expect to go back to initial server.
=> Her, I move the uniqAddr info from global var, to Instance related storage. But still share by all Metrics structures used during that instance.


## 2- Healther and what to do when the listener (uniq resource) cannot start.
I was going toward Health plugin for consistency : it should warn when CoreDNS service is not correct.
But I agree that EXIT the process is another valid solution.
NOTE: having the monitoring check the Metrics endpoint along with Health endpoint can work too .. but that is inconsistent with Health definition.

## 3- the size of the change ... hum here I am not following too much. 
The point is to have the plugin working correctly.
Here I tried to fix the following issues I see in the current code:
- move away from the global var 
- change init uniqAddr in Setup- away from the go function that we are not sure when it will run (I mean nothing ensure it will run before the Start of the server)
- set status in UniqAddr to "done" only when listener is really started. Seems sanity. But also allow to have either an external check (Health) or retry (if we want ...)
- mutex on UniqAddr functions. is sanity. We are lucky so far this lib is used only at setup or event that are synced.

As the code is done and proposed now, isn't it better to checked-in ?

- the one that want to monitor on endpoint Metrics can still continue
- the issue #1868 is fixed with this code
- lib uniqAddr is thread-safe and dan be reused 

and one usage I may have (depends internal Infoblox) is ability to retry the Start of the Metrics from within an external plugin which is doing CLI+monitoring.


## What I expect from metrics for IB usage
- fix of #1868 (after several reload, Metrics lost ability to stop/start the listener according to Corefile)
- ability to know it Metrics failed to launch the Listener
- ability to retry (ability to know is similar to Healther). and the retry is described into RepairHealth in UT.

NOTE: Right now, the EXIT in Metrics in case of fail to listen would hurt IB use case. 
They need to move up to Go 1.10. I am pushing on it. But not done yet.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Plugin/Metric..." ]
>- fix of #1868 (after several reload, Metrics lost ability to stop/start the listener according to Corefile)
>- ability to know it Metrics failed to launch the Listener
>- ability to retry (ability to know is similar to Healther). and the retry is described into RepairHealth in UT.
>
>NOTE: Right now, the EXIT in Metrics in case of fail to listen would hurt IB use case.
>They need to move up to Go 1.10. I am pushing on it. But not done yet.

go 1.10 has a fix, so you should just use that. I understand that moving an org 
to a new version is hard, but we can't just add code because that is somehow 
easier for you right now.

Coming back to failing the reload - metrics plugin should probably call panic 
and let the - whatever is managing it - restart it. Fun thing is that panic are 
captured by the errors plugin \o/

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
[ Quoting <miek@miek.nl> in "Re: [coredns/coredns] Plugin/Metric..." ]
>Coming back to failing the reload - metrics plugin should probably 
>call panic and let the - whatever is managing it - restart it. Fun 
>thing is that panic are captured by the errors plugin \o/

or log.Fatalf that calls os.Exit

**Comment by fturib** [comment]
About go version, I asked IB people, and look like we have an issue besides this version.

> I have asked the QA and  Gowtham mentions that the issue is observed on all  3 platforms :  Baremetal,  OVA,  and NIOS.
>  
> We are using golang 1.10.x   version,  so I guess it’s not related to the go issue that we are thinking of.

**Comment by fturib** [comment]
I will remove this PR, until we have the root cause of Listener failing to listen when restart.
Seems not the known Golang issue .. then, what else ?

the 2 issues are still there, but maybe not the important part.
